File: scripts/firebase/extract_service_account.py
#!/usr/bin/env python3
import argparse
import json
import sys


def read_env_value(env_path, key):
    try:
        with open(env_path, "r", encoding="utf-8") as handle:
            for raw_line in handle:
                line = raw_line.strip()
                if not line or line.startswith("#"):
                    continue
                if not line.startswith(f"{key}="):
                    continue
                return line.split("=", 1)[1].strip()
    except FileNotFoundError:
        return None
    return None


def normalize_json(raw):
    value = raw.strip()
    if (value.startswith('"') and value.endswith('"')) or (value.startswith("'") and value.endswith("'")):
        value = value[1:-1]
    try:
        data = json.loads(value)
    except json.JSONDecodeError:
        try:
            data = json.loads(value.encode("utf-8").decode("unicode_escape"))
        except Exception as exc:
            raise ValueError(f"No se pudo parsear JSON: {exc}") from exc
    if isinstance(data, str):
        data = json.loads(data)
    return json.dumps(data, indent=2)


def main():
    parser = argparse.ArgumentParser(
        description="Extrae FIREBASE_SERVICE_ACCOUNT de un archivo .env y lo imprime como JSON."
    )
    parser.add_argument("--env-file", default=".env.local", help="Ruta al archivo .env")
    parser.add_argument("--key", default="FIREBASE_SERVICE_ACCOUNT", help="Nombre de la variable")
    parser.add_argument("--out", default="-", help="Archivo de salida o '-' para stdout")
    args = parser.parse_args()

    raw = read_env_value(args.env_file, args.key)
    if not raw:
        print(f"No se encontro {args.key} en {args.env_file}", file=sys.stderr)
        sys.exit(1)

    try:
        json_payload = normalize_json(raw)
    except ValueError as exc:
        print(str(exc), file=sys.stderr)
        sys.exit(1)

    if args.out == "-":
        print(json_payload)
    else:
        with open(args.out, "w", encoding="utf-8") as handle:
            handle.write(json_payload)
            handle.write("\n")


if __name__ == "__main__":
    main()

File: scripts/firebase/extract_service_account.js
#!/usr/bin/env node
const fs = require("fs");

function getArg(name, defaultValue) {
  const idx = process.argv.indexOf(name);
  if (idx === -1) return defaultValue;
  return process.argv[idx + 1] || defaultValue;
}

function readEnvValue(envPath, key) {
  if (!fs.existsSync(envPath)) return null;
  const lines = fs.readFileSync(envPath, "utf8").split(/\r?\n/);
  for (const rawLine of lines) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;
    if (!line.startsWith(`${key}=`)) continue;
    return line.split("=", 1)[0].length === key.length
      ? line.slice(key.length + 1)
      : null;
  }
  return null;
}

function decodeEscapes(input) {
  const escaped = input.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
  return JSON.parse(`"${escaped}"`);
}

function normalizeJson(raw) {
  let value = raw.trim();
  if (
    (value.startsWith('"') && value.endsWith('"')) ||
    (value.startsWith("'") && value.endsWith("'"))
  ) {
    value = value.slice(1, -1);
  }
  let parsed;
  try {
    parsed = JSON.parse(value);
  } catch (_err) {
    const decoded = decodeEscapes(value);
    parsed = JSON.parse(decoded);
  }
  if (typeof parsed === "string") {
    parsed = JSON.parse(parsed);
  }
  return JSON.stringify(parsed, null, 2);
}

const envFile = getArg("--env-file", ".env.local");
const key = getArg("--key", "FIREBASE_SERVICE_ACCOUNT");
const out = getArg("--out", "-");

const raw = readEnvValue(envFile, key);
if (!raw) {
  console.error(`No se encontro ${key} en ${envFile}`);
  process.exit(1);
}

let jsonPayload;
try {
  jsonPayload = normalizeJson(raw);
} catch (err) {
  console.error(`No se pudo parsear JSON: ${err.message}`);
  process.exit(1);
}

if (out === "-") {
  console.log(jsonPayload);
} else {
  fs.writeFileSync(out, `${jsonPayload}\n`, "utf8");
}
